spel/scripts/amigen8-build.sh

```diff
@@ -0,0 +1,412 @@
+   else
+      DISKSETUPCMD+="-p ${AMIGENSTORLAY} "
+   fi
+
+   # Set LVM2 or bare disk-formatting
+   if [[ ${AMIGENVGNAME} != "UNDEF" ]]
+   then
+      DISKSETUPCMD+="-v ${AMIGENVGNAME} "
+   elif [[ ${AMIGENROOTNM} != "UNDEF" ]]
+   then
+      DISKSETUPCMD+="-r ${AMIGENROOTNM} "
+   fi
+
+   # Set device to carve
+   if [[ ${AMIGENBUILDDEV} == "UNDEF" ]]
+   then
+      err_exit "Failed to define device to partition"
+   else
+      DISKSETUPCMD+="-d ${AMIGENBUILDDEV}"
+   fi
+
+   # Return command-string for disk-setup script
+   echo "${DISKSETUPCMD}"
+}
+
+# Pick options for the OS-install command
+function ComposeOSpkgString {
+   local OSPACKAGESTRING
+
+   OSPACKAGESTRING="OSpackages.sh "
+
+   # Set location for chroot-env
+   if [[ ${AMIGENCHROOT} == "/mnt/ec2-root" ]]
+   then
+      err_exit "Using default chroot-env location [${AMIGENCHROOT}]" NONE
+   else
+      OSPACKAGESTRING+="-m ${AMIGENCHROOT} "
+   fi
+
+   # Pick custom yum repos
+   if [[ ${AMIGENREPOS} == "UNDEF" ]]
+   then
+      err_exit "Using script-default yum repos" NONE
+   else
+      OSPACKAGESTRING+="-a ${AMIGENREPOS} "
+   fi
+
+   # Custom repo-def RPMs to install
+   if [[ ${AMIGENREPOSRC} == "UNDEF" ]]
+   then
+      err_exit "Installing no custom repo-config RPMs" NONE
+   else
+      OSPACKAGESTRING+="-r ${AMIGENREPOSRC}"
+   fi
+
+   # Return command-string for OS-script
+   echo "${OSPACKAGESTRING}"
+}
+
+function PostBuildString {
+   local POSTBUILDCMD
+
+   POSTBUILDCMD="PostBuild.sh "
+
+   # Set the filesystem-type to use for OS filesystems
+   if [[ ${AMIGENFSTYPE} == "xfs" ]]
+   then
+      err_exit "Using default fstype [xfs] for boot filesysems" NONE
+   fi
+   POSTBUILDCMD+="-f ${AMIGENFSTYPE} "
+
+   # Set location for chroot-env
+   if [[ ${AMIGENCHROOT} == "/mnt/ec2-root" ]]
+   then
+      err_exit "Using default chroot-env location [${AMIGENCHROOT}]" NONE
+   else
+      POSTBUILDCMD+="-m ${AMIGENCHROOT} "
+   fi
+
+   # Set AMI starting time-zone
+   if [[ ${AMIGENTIMEZONE} == "UTC" ]]
+   then
+      err_exit "Using default AMI timezone [${AMIGENCHROOT}]" NONE
+   else
+      POSTBUILDCMD+="-z ${AMIGENTIMEZONE} "
+   fi
+
+   # Return command-string for OS-script
+   echo "${POSTBUILDCMD}"
+}
+
+# Disable strict hostkey checking
+function DisableStrictHostCheck {
+   local HOSTVAL
+
+   if [[ ${1:-} == '' ]]
+   then
+      err_exit "No connect-string passed to function [${0}]"
+   else
+      HOSTVAL="$( sed -e 's/^.*@//' -e 's/:.*$//' <<< "${1}" )"
+   fi
+
+   # Git host-target parameters
+   err_exit "Disabling SSH's strict hostkey checking for ${HOSTVAL}" NONE
+   (
+      printf "Host %s\n" "${HOSTVAL}"
+      printf "  Hostname %s\n" "${HOSTVAL}"
+      printf "  StrictHostKeyChecking off\n"
+   ) >> "${HOME}/.ssh/config" || \
+   err_exit "Failed disabling SSH's strict hostkey checking"
+}
+
+
+
+##########################
+## Main program section ##
+##########################
+
```

Comment:
still need to account here for the [setup of the buildhost](https://github.com/plus3it/spel/blob/master/spel/scripts/amigen7-build.sh#L151-L174)

Reply:
Err... Not following. Buildhost setup already happens in [this block](https://github.com/ferricoxide/spel/blob/EL8/spel/scripts/amigen8-build.sh#L359-L393)?

Reply:
that's only part of the setup. the part i highlighted in the link to the amigen7 script is to account for cross-environment deltas.

Reply:
Ok, so, why I'm confused is, the logic you seem to be pointing to is already baked into AMIgen8's `OSpackages.sh` script, I believe (mostly its [`PrepChroot` function](https://github.com/plus3it/AMIgen8/blob/master/OSpackages.sh#L153-L220)).

Reply:
The difference is all that runs inside the chroot, essentially configuring the resulting ami. What I'm referring to is working against the build host itself, not the chroot.

Reply:
Ok, I've got a case of the stupids, today, because I'm _really_ not able to track what you're saying. Best I can suss out is:

- Looking at lines [155-156](https://github.com/plus3it/spel/blob/master/spel/scripts/amigen7-build.sh#L154-L156) of the linked-to section, that's already taken care of in the userData file's `#cloud-config` section
- Looking at lines [158-167](https://github.com/plus3it/spel/blob/master/spel/scripts/amigen7-build.sh#L158-L167), should be a function of the bootstrap AMI's initial configuration (i.e., the private networks' source-AMIs already have the repos used for the relevant network configured)?
- Looking at lines [169-174](https://github.com/plus3it/spel/blob/master/spel/scripts/amigen7-build.sh#L169-L174), there's currently not an analogous need?
- Looking at [176-L180](https://github.com/plus3it/spel/blob/master/spel/scripts/amigen7-build.sh#L176-L180) that's in the `amigen8-build.sh` script's `DisableStrictHostCheck` function

Reply:
> Looking at lines 155-156  of the linked-to section, that's already taken care of in the userData file's #cloud-config section

Userdata should not be doing anything more than the absolute minimum. We want packer to connect as quickly as possible, so logs start streaming earlier.

> Looking at lines 158-167 , should be a function of the bootstrap AMI's initial configuration (i.e., the private networks' source-AMIs already have the repos used for the relevant network configured)?

Unfortunately, the source-AMIs rather frequently have some kind of problem due to some kind of change in the environment, and it takes several days to figure it out and generate new ones. We've found that what we specify as custom repos in the chroot for a new AMI, are the same as what we need in the source-AMI builder. So we now install these things into the builder, which gives us a lot more flexibility is source-AMI selection and use.

> Looking at lines 169-174 , there's currently not an analogous need?

We use these variables in very different ways when in isolated environments. Similar to the above, what we specify as EXTRARPMS for the chroot, we've found are also what we need for the source-AMI builder. This is a big part of why we added the generic "extra rpm" feature to amigen6 and amigen7. It gives us a lot flexibility to tailor the build process for different environments.

But we're still a ways off from publishing EL8 AMIs. We can wait on addressing isolated network concerns. We can try to avoid any promises about EL8 availability until we've actually tested it there and come back to re-address.

Reply:
> Userdata should not be doing anything more than the absolute minimum. We want packer to connect as quickly as possible, so logs start streaming earlier.

Ok... Was confused: the userdata script was already doing a `yum upgrade`. I assumed from that that adding the three missing package was appropriate there:

~~~
$ git diff upstream/master..HEAD -- userdata/tmpfsroot-el7.cloud
diff --git a/spel/userdata/tmpfsroot-el7.cloud b/spel/userdata/tmpfsroot-el7.cloud
index 6ee1c7c..f2cf651 100644
--- a/spel/userdata/tmpfsroot-el7.cloud
+++ b/spel/userdata/tmpfsroot-el7.cloud
@@ -26,6 +26,11 @@ bootcmd:

 package_upgrade: true

+packages:
+  - git
+  - lvm2
+  - unzip
+
 --===============3585321300151562773==
~~~
Didn't figure timing was a concern since the time to install the further three, small packages is significantly shorter than the amount of time to apply a full system update – especially if the bootstrap AMI is more than, say 60 days old. At any rate, I'll move them from the userdata to the `amigen7-build.sh` script.

> Unfortunately, the source-AMIs rather frequently have some kind of problem due to some kind of change in the environment, and it takes several days to figure it out and generate new ones. 

Only recent one I was aware of was when the Ops group botched the December repo-sync and forgot the groups metadata (and, for EL7, we didn't have a workaround for that, at any rate, until a couple months later): have there been other issues that weren't reported?

> We've found that what we specify as custom repos in the chroot for a new AMI, are the same as what we need in the source-AMI builder. So we now install these things into the builder, which gives us a lot more flexibility is source-AMI selection and use.

Ah, yeah, now I remember: that was a RHEL issue w/ respect to the our custom repos (since we use vendor-AMIs to bootstrap RHEL rather than using a private bootstrap AMI).

Reply:
Alright, fcdecf1 moves the builder-package logic from userData to buildhost setup-script.

We're waiting on the disconnected networks one till we have access to those again, correct (i.e., this conversation is resolved by the linked commit)?

Reply:
> Ok... Was confused: the userdata script was already doing a yum upgrade.

I believe, now, that was a mistake. Wasn't thinking about the consequence to packer connecting and beginning the log-stream [2.5 years ago when that change was made0(https://github.com/plus3it/spel/pull/77)... The yum update helps address issues with changes to the repo definitions/access, particularly when a minor update is involved.

Reply:
> have there been other issues

There has been issue after issue after issue. Eventually, got fed up with the dependency on fixing the source-ami, and realized that we could address quite a few problems just by setting up the builder-host with most-all of the same things that we push into the chroot. Since the builder-host is discarded anyway, it did not matter if some of those things ended up being unnecessary, as long as others provided the required flexibility to work around issues that would otherwise require a whole new source-ami.

Reply:
> We're waiting on the disconnected networks one till we have access to those again

It will burn some, knowing already that this will fail and require addressing the same problems that are already fixed on the el7 side. But, since I won't be the one dealing with it, I can swallow the bile. 😉

Reply:
Mostly, I'm waiting on _that_ to see if, once it presents itself, I want to use a different mechanism for solving the problem (rather than immediately cut-n-pasting prior kludges).